File: projects/velvet-forge-m8k4/src/VelvetForge/Core/LdPlayer.cs
namespace VelvetForge.Core;

public sealed record LdInstanceInfo(
  int Index,
  string Title,
  int TopHwnd,
  int BindHwnd,
  int AndroidStarted,
  int Pid,
  int VboxPid,
  int Width,
  int Height,
  int Dpi);

public sealed class LdPlayer(FarmSettings settings, Action<string>? log = null)
{
  private readonly FarmSettings _s = settings;
  private string? _home;
  private int? _adbPort;

  public string Home => _home ??= ResolveHome();
  public string LdConsole => Path.Combine(Home, "ldconsole.exe");
  public string AdbBin => Path.Combine(Home, "adb.exe");
  public int Index => _s.LdPlayerIndex;

  private void Log(string msg)
  {
    log?.Invoke(msg);
    Console.WriteLine(msg);
  }

  public string ResolveHome()
  {
    var candidates = new List<string>();
    if (!string.IsNullOrWhiteSpace(_s.LdPlayerHome)) candidates.Add(_s.LdPlayerHome);
    var env = Environment.GetEnvironmentVariable("LDPLAYER_HOME");
    if (!string.IsNullOrWhiteSpace(env)) candidates.Add(env);
    candidates.AddRange(
    [
      @"D:\LDPlayer\LDPlayer9",
      @"C:\Program Files\LDPlayer\LDPlayer9",
      @"C:\Program Files (x86)\LDPlayer\LDPlayer9",
    ]);
    foreach (var c in candidates.Distinct())
    {
      if (File.Exists(Path.Combine(c, "ldconsole.exe"))) return c;
    }
    throw new FileNotFoundException("LDPlayer not found. Set ldplayerHome in settings.json");
  }

  public int ResolveAdbPort()
  {
    if (_adbPort is not null) return _adbPort.Value;
    var configPath = Path.Combine(Home, "config", "ld.config");
    if (File.Exists(configPath))
    {
      foreach (var line in File.ReadAllLines(configPath))
      {
        if (line.StartsWith("adb_port=", StringComparison.OrdinalIgnoreCase) &&
            int.TryParse(line["adb_port=".Length..], out var port))
        {
          _adbPort = port;
          return port;
        }
      }
    }
    _adbPort = 5555 + Index * 2;
    return _adbPort.Value;
  }

  public string Run(params string[] args)
  {
    var psi = new System.Diagnostics.ProcessStartInfo(LdConsole, string.Join(" ", args))
    {
      RedirectStandardOutput = true,
      RedirectStandardError = true,
      UseShellExecute = false,
      CreateNoWindow = true,
    };
    using var p = System.Diagnostics.Process.Start(psi)!;
    var stdout = p.StandardOutput.ReadToEnd();
    var stderr = p.StandardError.ReadToEnd();
    p.WaitForExit();
    return (stdout + stderr).Trim();
  }

  public LdInstanceInfo? GetInstanceInfo()
  {
    var lines = Run("list2").Split('\n', StringSplitOptions.RemoveEmptyEntries);
    foreach (var line in lines)
    {
      if (!line.StartsWith($"{Index},", StringComparison.Ordinal)) continue;
      var parts = line.Split(',');
      if (parts.Length < 10) return null;
      return new LdInstanceInfo(
        int.Parse(parts[0]),
        parts[1],
        int.Parse(parts[2]),
        int.Parse(parts[3]),
        int.Parse(parts[4]),
        int.Parse(parts[5]),
        int.Parse(parts[6]),
        int.Parse(parts[7]),
        int.Parse(parts[8]),
        int.Parse(parts[9]));
    }
    return null;
  }

  public bool IsRunning()
  {
    var outText = Run("isrunning", "--index", Index.ToString());
    return outText.Contains("running", StringComparison.OrdinalIgnoreCase);
  }

  public void Launch()
  {
    if (IsRunning())
    {
      Log($"ldplayer: already running index={Index}");
      return;
    }
    Log($"ldplayer: launching index={Index}");
    Run("launch", "--index", Index.ToString());
  }

  public void Reboot()
  {
    Log($"ldplayer: reboot index={Index}");
    Run("reboot", "--index", Index.ToString());
  }

  public void ModifyRoot() => Run("modify", "--index", Index.ToString(), "--root", "1");

  public void RunApp(string? pkg = null)
  {
    pkg ??= _s.MlbbPackage;
    Run("runapp", "--index", Index.ToString(), "--packagename", pkg);
  }

  public void KillApp(string? pkg = null)
  {
    pkg ??= _s.MlbbPackage;
    Run("killapp", "--index", Index.ToString(), "--packagename", pkg);
  }

  public bool Screenshot(string outPath)
  {
    FarmPaths.EnsureDataDirs();
    Directory.CreateDirectory(Path.GetDirectoryName(outPath)!);
    Run("scan", "--index", Index.ToString(), "--file", outPath);
    return File.Exists(outPath);
  }

  public string AdbCommand(string command) =>
    Run("adb", "--index", Index.ToString(), "--command", command);

  public void SendKeyboard(string key) =>
    Run("action", "--index", Index.ToString(), "--key", "call.keyboard", "--value", key);

  public bool IsPackageInstalled(string package)
  {
    var output = AdbCommand($"shell pm path {package}");
    return output.Contains("package:", StringComparison.OrdinalIgnoreCase);
  }
}

File: projects/wild-pulse-n8k3/wild_pulse/ldplayer.py
from __future__ import annotations

import json
import subprocess
import time
from pathlib import Path

from .settings import Settings


class LdPlayer:
    """LDPlayer ldconsole wrapper."""

    def __init__(self, settings: Settings, log=print):
        self.s = settings
        self.log = log
        self.home = self._resolve_home()
        self.ldconsole = self.home / "ldconsole.exe"
        self.adb = self.home / "adb.exe"
        self.index = settings.ldplayer_index

    def _resolve_home(self) -> Path:
        candidates = [
            Path(self.s.ldplayer_home),
            Path(__import__("os").environ.get("LDPLAYER_HOME", "")),
            Path(r"D:\LDPlayer\LDPlayer9"),
            Path(r"C:\Program Files\LDPlayer\LDPlayer9"),
        ]
        for c in candidates:
            if c and (c / "ldconsole.exe").exists():
                return c
        raise FileNotFoundError("LDPlayer not found — set ldplayerHome in settings.json")

    def run(self, *args: str) -> str:
        cmd = [str(self.ldconsole), *args]
        p = subprocess.run(cmd, capture_output=True, text=True)
        return (p.stdout + p.stderr).strip()

    def is_running(self) -> bool:
        return "running" in self.run("isrunning", "--index", str(self.index)).lower()

    def launch(self) -> None:
        if self.is_running():
            self.log(f"ldplayer: already running index={self.index}")
            return
        self.log(f"ldplayer: launching index={self.index}")
        self.run("launch", "--index", str(self.index))

    def reboot(self) -> None:
        self.log(f"ldplayer: reboot index={self.index}")
        self.run("reboot", "--index", str(self.index))

    def modify_root(self) -> None:
        self.run("modify", "--index", str(self.index), "--root", "1")

    def run_app(self, package: str | None = None) -> None:
        pkg = package or self.s.mlbb_package
        self.run("runapp", "--index", str(self.index), "--packagename", pkg)

    def kill_app(self, package: str | None = None) -> None:
        pkg = package or self.s.mlbb_package
        self.run("killapp", "--index", str(self.index), "--packagename", pkg)

    def install_app(self, package: str | None = None) -> str:
        pkg = package or self.s.mlbb_package
        return self.run("installapp", "--index", str(self.index), "--packagename", pkg)

    def list2(self) -> str:
        return self.run("list2")

    def adb_shell(self, command: str) -> str:
        return self.run("adb", "--index", str(self.index), "--command", command)

    def operaterecord(self, operations: list[dict], loop_times: int = 1) -> None:
        payload = {
            "operations": operations,
            "recordInfo": {
                "loopType": 2,
                "loopTimes": loop_times,
                "circleDuration": 200,
                "loopInterval": 0,
                "loopDuration": 0,
                "accelerateTimes": 1,
                "accelerateTimesEx": 1,
            },
        }
        content = json.dumps(payload, separators=(",", ":"))
        self.run("operaterecord", "--index", str(self.index), "--content", content)

    def wait_boot(self, timeout: int = 120) -> bool:
        deadline = time.time() + timeout
        while time.time() < deadline:
            if self.is_running():
                time.sleep(3)
                return True
            time.sleep(2)
        return False
